src/v/cluster/archival_metadata_stm.cc

```diff
@@ -650,6 +785,8 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
         co_return result.error();
     }
 
+    _last_replicated_offset = result.value().last_offset;
```

Comment:
I think it still doesn't solve the issue. Consider the following scenario:

* old archiver calls `replicate` and then `ssx::with_timeout_abortable`. The `_last_replicated_offset` contains old value;
* something stops the archiver midterm and `ssx::with_timeout_abortable` exits but the replication operation continues;
* new archiver calls `sync` and uses old `_last_replicated_offset` value;

Reply:
Yea this is a good point. Replication that is attempted and appear to fail (eg because of a timeout) may still succeed, and I don't think we have a bulletproof way of knowing to wait for it.

I wondered briefly if we could cache the current committed offset before calling replicate, but it won't work given other calls to replicate may race with us.

Reply:
Had a chat with Evgeny about this. Our idea was to chain the update to `_last_replicated_offset` to the `replicate` future. That works, but thinking about it some more, I see another issue: the `replicate` fiber might not have finished by the time the archiver restarts so the new instance doesn't know it's supposed to wait.

In theory we could keep the `future` for the latest `replicate` in the STM and always wait on it in our `sync` implementation. Issue is that there may be multiple replication fibres in flight (from the segment merger and scrubber) and I'm not sure about the semantics of the "final" replicate. I've asked the replication team.